src/main/java/org/jenkinsci/plugins/pubsub/listeners/SyncQueueListener.java

```diff
@@ -108,7 +109,7 @@
                                     // the "try later" queue ensures that it will be left for a little while
                                     // if the main queue is empty i.e. we avoid a tight loop here.
                                     // See top level comments.
-                                    tryLaterQueueTaskLeftQueue.put(leftItem);
```

Comment:
maybe we could save with one single line change? 
What we want to know at this stage. Is this leftItem still in the queue? because here we probably put in a different queue an item which is already not in the queue anymore.
why not simply testing if still in queue and discard or not
```
if (Queue.getInstance().getItem(leftItem.getId())!=null) {
  tryLaterQueueTaskLeftQueue.put(leftItem);
}
```
that would save few changes with a single line change ( (ok 2 lines)

Reply:
My changes clean the queue because it checks in line 93 whether the item, now read from the real queue, is null. If it is, it's not added to the queue anymore.

Reply:
My goal with the PR is, in addition to remove the elements from the queue when they're not valid, remove the class fields holding an Item, which I think is a bad practice, as I read from a @jglick's comment in a related issue.

Reply:
by changing the structure of the stored data, your test will only be able to check if the item id is in the queue or not 
Here the logic is to publish a message once the job/task has been completed. 
If not we do not publish anymore the message saying saying task or job has been completed (i.e BlueOcean will not be informed anymore of this event).
The goal of this loop is to check is the job/task has been completed (see https://github.com/jenkinsci/pubsub-light-plugin/blob/b8388a135fe57b39a98464844f872336bad336c0/src/main/java/org/jenkinsci/plugins/pubsub/listeners/SyncQueueListener.java#L95) . By storing only the id we cannot verify anymore this.
With your change here (https://github.com/jenkinsci/pubsub-light-plugin/pull/42/files#diff-db2b174f8d7cfc36b1d3e2dd17ab41cededf617835e55390ec8fa5a3433fb7cbR92), if the item is in the queue it is definitely not done because still in the queue. So you can never have this test positive and you will never publish the message.

Reply:
The code after the change is still checking `future.isDone()` and furthermore, doing so with the current object, not a potentially ghost one stored in the queue. So I think it works as expected.

Reply:
it is potentially a ghost as it is not anymore really in the Queue but it is stored in another field `leftItems` of the Queue for only a certain amount of time (5 minutes) and this temporary cache can be clear at anytime by a script by executing the method `Queue.getInstance().clearLeftItems()`. Yes it works because of this cache (that's probably another discussion as why a queue still keep reference to elements not anymore in the queue which doesn't look right but this is probably made by design.... but still a bit ugly...) but definitely it's fragile to rely on this "cache" whereas you don't need as we had direct access to the `Item` without need to query the Queue. 
The original problem is the queue not cleared...
Anyway it works because of this cache implementation detail....

Reply:
We can also just change the references to `WeakReference`s to have them nulled after the Queue.Item or the Item is removed.

Anyway, I would like to avoid this plugin showing up in the heap dumps as a big memory consumer, because it tricks people analyzing them. If we merge your PR, I would like to remove the Jenkins local stored variable and any unneeded Item or Queue.Item.

Reply:
oh I would prefer avoid something even more complex with `WeakReference`. I do see not what will be the gain with such complexity. Because as long as we remove them from the `BlockingQueue` there are still referenced by the cache in the `Queue` instance for a minimum of 5 minutes :) 

Again the problem is not what it is stored in the `BlockingQueue` fields but the `BlockingQueue` fields not cleared, OOME is the symptom the cause is not clearing those `BlockingQueue` fields.
You will not save any memory here. If we clear correctly the `BlockingQueue` fields we just have reference to other `Item` already stored in the `Queue` so we do not consume more memory which are by the way stored longer than necessary because of this cache in `Queue`.
Yes this PR fix the problem but I see a lot of changes for no real memory gain for something which can be saved with 3 lines of code change (https://github.com/jenkinsci/pubsub-light-plugin/pull/43/files).
I know I'm a real pain but I usually prefer small changes when they solve a problem ;) 
Yes I will be happy to see some memory improvements in another PR and I'm curious to see that.